src/sdk/namespace_fs.js

```diff
@@ -492,23 +494,94 @@ class NamespaceFS {
+        if (params.copy_source) {
+            const source_file_path = path.join(this.bucket_path, params.copy_source.key);
+            try {
+                const same_inode = await this._is_same_inode(fs_account_config, source_file_path, file_path);
+                if (same_inode) return same_inode;
+                // Doing a hard link.
+                await nb_native().fs.link(fs_account_config, source_file_path, upload_path);
+            } catch (e) {
+                try {
+                    console.warn('NamespaceFS: COPY using link failed with:', e);
+                    await this._copy_stream(source_file_path, upload_path, fs_account_config);
+                } catch (err) {
+                    throw this._translate_object_error_codes(err);
+                }
+            }
+        } else {
+            try {
+                await this._upload_stream(params.source_stream, upload_path, fs_account_config);
+            } catch (err) {
+                throw this._translate_object_error_codes(err);
+            }
+        }
+        // TODO use file xattr to store md5_b64 xattr, etc.
+        const stat = await nb_native().fs.stat(fs_account_config, upload_path);
+        await nb_native().fs.rename(fs_account_config, upload_path, file_path);
+        return { etag: this._get_etag(stat) };
+    }
 
-            await Promise.all([this._make_path_dirs(file_path, fs_account_config), this._make_path_dirs(upload_path, fs_account_config)]);
-            await this._upload_stream(params.source_stream, upload_path, fs_account_config);
-            // TODO use file xattr to store md5_b64 xattr, etc.
-            const stat = await nb_native().fs.stat(fs_account_config, upload_path);
-            await nb_native().fs.rename(fs_account_config, upload_path, file_path);
-            return { etag: this._get_etag(stat) };
-        } catch (err) {
-            throw this._translate_object_error_codes(err);
+    // Comparing both device and inode number (st_dev and st_ino returned by stat) 
+    // will tell you whether two different file names refer to the same thing.
+    // If so, we will return the etag of the file_path
+    async _is_same_inode(fs_account_config, source_file_path, file_path) {
+        try {
+            const file_path_stat = await nb_native().fs.stat(fs_account_config, file_path);
+            const file_path_inode = file_path_stat.ino.toString();
+            const file_path_device = file_path_stat.dev.toString();
+            const source_file_stat = await nb_native().fs.stat(fs_account_config, source_file_path);
+            const source_file_inode = source_file_stat.ino.toString();
+            const source_file_device = source_file_stat.dev.toString();
+            if (file_path_inode === source_file_inode && file_path_device === source_file_device) {
+                return { etag: this._get_etag(file_path_stat) };
+            }
+        } catch (e) {
+            return false;
```

Comment:
Why do we ignore failures and just assume that if the stat failed the inode is not the same?

Reply:
This function is to decide if we should try to link or not.
In the event of an already exist copy, which is with the same name and the same inode, linking will work but `renameat2` will fail and we will leave with more than 2 links (we will have an additional one in the temp dir).
So we are checking if we would link or not, and if we fail for any reason, we will try to link.